### PR TITLE
typings: withSize add onSize callback

### PR DESCRIPTION
**Pull request by beckend**

### CODE REVIEW

**Comment by beckend** [comment]
@ctrlplusb can you review?

File: react-sizeme.d.ts [hunk 1]
```diff
@@ -3,29 +3,42 @@ import { Component, ComponentType, ReactNode, ReactElement } from 'react'
...
 }
 
 export interface SizeMeOptions {
-  monitorWidth?: boolean
-  monitorHeight?: boolean
-  monitorPosition?: boolean
-  refreshRate?: number
-  refreshMode?: 'throttle' | 'debounce'
-  noPlaceholder?: boolean
+  readonly monitorWidth?: boolean
```
Comment (ctrlplusb) [inline react-sizeme.d.ts]: Not sure if there is any value marking these as read-only as they won't get exposed to the user. These configuration values are inputs.

File: react-sizeme.d.ts [hunk 2]
```diff
@@ -3,29 +3,42 @@ import { Component, ComponentType, ReactNode, ReactElement } from 'react'
 type Omit<T, K> = Pick<T, Exclude<keyof T, K>>
 
 export interface SizeMeProps {
-  size: {
-    width: number | null
-    height: number | null
+  readonly size: {
```
Comment (ctrlplusb) [inline react-sizeme.d.ts]: Nice, good touch. 👍

File: react-sizeme.d.ts [hunk 3]
```diff
@@ -3,29 +3,42 @@ import { Component, ComponentType, ReactNode, ReactElement } from 'react'
...
+  readonly monitorHeight?: boolean
+  readonly monitorPosition?: boolean
+  readonly refreshRate?: number
+  readonly refreshMode?: 'throttle' | 'debounce'
+  readonly noPlaceholder?: boolean
 }
 
 export interface SizeMeRenderProps extends SizeMeOptions {
-  children: (props: SizeMeProps) => ReactElement
+  readonly children: (props: SizeMeProps) => ReactElement
```
Comment (ctrlplusb) [inline react-sizeme.d.ts]: Same here, this is user provided, and won't get exposed in a manner that the user could modify it, so the `readonly` isn't really useful in this case.

File: react-sizeme.d.ts [hunk 4]
```diff
@@ -3,29 +3,42 @@ import { Component, ComponentType, ReactNode, ReactElement } from 'react'
...
 }
 
 export class SizeMe extends Component<SizeMeRenderProps> {}
 
+export type WithSizeOnSizeArgs = [
```
Comment (ctrlplusb) [inline react-sizeme.d.ts]: Could we rename this to `Size` and perhaps reuse the type within `SizeMeProps`?

File: react-sizeme.d.ts [hunk 5]
```diff
@@ -3,29 +3,42 @@ import { Component, ComponentType, ReactNode, ReactElement } from 'react'
...
+  {
+    readonly height: number | null
+    readonly width: number | null
+  }
+]
+
+export type WithSizeOnSizeCallback = (...x: WithSizeOnSizeArgs) => void
```
Comment (ctrlplusb) [inline react-sizeme.d.ts]: You are spreading an array/tuple that contains a single item. Perhaps consider simplifying this (after the `Size` type refactor I recommended above) to the following:

```typescript
export type OnSizeCallback = (size: Size) => void;
```

File: react-sizeme.d.ts [hunk 6]
```diff
@@ -3,29 +3,42 @@ import { Component, ComponentType, ReactNode, ReactElement } from 'react'
...
+  {
+    readonly height: number | null
+    readonly width: number | null
+  }
+]
+
+export type WithSizeOnSizeCallback = (...x: WithSizeOnSizeArgs) => void 
+
+export interface WithSizeProps { 
+  readonly onSize?: WithSizeOnSizeCallback
```
Comment (ctrlplusb) [inline react-sizeme.d.ts]: Same here regarding the `readonly` not really being useful as it's an input that doesn't get exposed to user again.

File: src/__tests__/typescript/hoc.tsx [hunk 7]
```diff
@@ -22,4 +22,8 @@ function MyComponent({ id, size }: MyComponentProps) {
 
 const SizedMyComponent = withSize()(MyComponent)
 
-const foo = <SizedMyComponent id={1} />
+const onSize: WithSizeOnSizeCallback = ({ height, width }) => {
+  Boolean(height && width)
```
Comment (ctrlplusb) [inline src/__tests__/typescript/hoc.tsx]: Could you replace the body of this with the following:

```typescript
  if (width) {
    const foo = width + 1	    const foo = width + 1
  }	  }
  if (height) {	  if (height) {
    const foo = height + 1	    const foo = height + 1
  }	  }
  // typings:expect-error	  // typings:expect-error
  const h1 = height + 1	  const h1 = height + 1
  // typings:expect-error	  // typings:expect-error
  const w1 = width + 1	  const w1 = width + 1
```

This helps guard that width/height are nullable numbers.

**Comment by ctrlplusb** [comment]
@beckend reviewed :)

**Comment by beckend** [comment]
@ctrlplusb sorry for the delay, life and stuff...
Anyway did the adjustments, I personally however would never allow argument mutation by default even if it is not being used, but I also understand that it is also trivial in this case.

**Comment by beckend** [comment]
The PR was done, however not action was taken.

**Comment by ctrlplusb** [comment]
Hey @beckend 

I still intend to merge this.

I just need to do some more planning around it.

Please leave open.

Thanks

**Comment by beckend** [comment]
Got it, just wanted to do some housecleaning.

**Pull request merged**